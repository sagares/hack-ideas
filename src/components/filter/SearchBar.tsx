import React from "react";

type SearchBarProps = {
  searchText: string;
  onFilter: (searchText: string) => void;
}

export const SearchBar = ({ searchText, onFilter }: SearchBarProps) => {
  const handleSearch = (e: React.FormEvent<HTMLInputElement>) => {
    onFilter(e.currentTarget.value);
  };

  const handleClear = () => {
    onFilter("");
  };

  return (
    <nav>
      <div className="nav-wrapper">
        <form>
          <div className="input-field grey lighten-4">
            <input
              id="search"
              type="search"
              value={searchText}
              onChange={handleSearch}
            />
            <label className="label-icon" htmlFor="search">
              <i className="material-icons">search</i>
            </label>
            <i className="clear-icon material-icons" onClick={handleClear}>
              close
            </i>
          </div>
        </form>
      </div>
    </nav>
  );
};
