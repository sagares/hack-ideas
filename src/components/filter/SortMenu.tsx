import React, { useEffect } from "react";
import M from "materialize-css";

type SortMenuProps = {
  sortBy: string;
  onSort: (sortBy: string) => void;
};

export const SortMenu = ({ sortBy, onSort }: SortMenuProps) => {
  useEffect(() => {
    var dropdowns = document.querySelectorAll(".dropdown-trigger");
    M.Dropdown.init(dropdowns, {});
  }, []);

  return (
    <React.Fragment>
      <a
        className="dropdown-trigger btn-floating btn-large"
        data-target="dropdown1"
      >
        <i className="large material-icons">sort</i>
      </a>

      <ul id="dropdown1" className="dropdown-content">
        <li className={`${sortBy === "submittedOn" ? "active" : ""}`}>
          <a onClick={() => onSort("submittedOn")}>
            <i className="material-icons">access_time</i>By Created On
          </a>
        </li>
        <li className={`${sortBy === "upvotes" ? "active" : ""}`}>
          <a onClick={() => onSort("upvotes")}>
            <i className="material-icons">thumb_up</i>By Upvotes
          </a>
        </li>
        <li className="divider" tabIndex={-1}></li>
        <li>
          <a onClick={() => onSort("")}>
            <i className="material-icons">cached</i>Reset
          </a>
        </li>
      </ul>
    </React.Fragment>
  );
};
