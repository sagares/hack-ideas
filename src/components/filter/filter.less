@import "../../colors.less";

.hack-ideas--filter {
    align-items: center;
    display: flex;
    justify-content: space-between;
    padding: 0.7em;
    margin-bottom: 0;

    .search-bar {
      .nav-wrapper {
        .input-field {
          &.grey {
            label {
              i {
                color: @body-text;
              }
            }
          }
        }
      }
      @media only screen and (min-width: 601px) {
        nav {
          .height50();
          .nav-wrapper {
            i {
              .height50();
            }
          }
        }
      }
    }
    .filter-dropdown {
      display: flex;
      margin: 0 !important;
      position: relative;

      .btn-large {
        background-color: @emphasis-clr;
        i {
          color: @dark-clr;
        }
      }
      .dropdown-content {
        height: auto !important;
        width: auto !important;
        li {
          > a {
            color: @emphasis-clr;
            display: flex;
          }
        }
      }
    }
  }

  .height50() {
    height: 50px;
    line-height: 50px;
  }