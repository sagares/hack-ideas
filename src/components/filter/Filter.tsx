import React from "react";
import { FilterProps } from "../../HackIdea.types";
import { SearchBar } from ".";
import { SortMenu } from "./SortMenu";

import "./filter.less";

const Filter = ({ searchText, onFilter, sortBy, onSort }: FilterProps) => {
  return (
    <div className="hack-ideas--filter row">
      <div className="search-bar col s12 m12 l4">
        <SearchBar searchText={searchText} onFilter={onFilter} />
      </div>
      <div className="filter-dropdown col s4 m2 l2">
        <SortMenu sortBy={sortBy} onSort={onSort} />
      </div>
    </div>
  );
};

export default Filter;
