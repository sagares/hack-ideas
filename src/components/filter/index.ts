import { SearchBar } from "./SearchBar";
import { SortMenu } from "./SortMenu";
import Filter from "./Filter";

export default Filter;
export { SearchBar, SortMenu }