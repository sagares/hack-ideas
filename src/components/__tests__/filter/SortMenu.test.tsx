/**
 * @jest-environment jsdom
 */
import React from "react";
import { fireEvent, render } from "@testing-library/react";
import { SortMenu } from "../../filter";

describe("Sort Menu Component", () => {
  let element, component, onSortSpy;

  beforeEach(() => {
    onSortSpy = jest.fn();

    element = document.createElement("div");

    component = render(
      <SortMenu sortBy="" onSort={onSortSpy} orderAsc={true} />,
      {
        container: document.body.appendChild(element),
      }
    );
  });

  afterEach(() => {
    document.body.removeChild(element);
  });

  test("renders", () => {
    expect(component.container).toMatchSnapshot();
  });

  test("sorts by submitted on", () => {
    const submittedOn = element.querySelector(
      ".dropdown-content li:first-child a"
    );
    fireEvent.click(submittedOn);
    expect(onSortSpy).toHaveBeenCalledWith("submittedOn");
  });

  test("sorts by unvotes count", () => {
    const upvotes = element.querySelector(
      ".dropdown-content li:nth-child(2) a"
    );
    fireEvent.click(upvotes);
    expect(onSortSpy).toHaveBeenCalledWith("upvotes");
  });

  test("rest sorting", () => {
    const upvotes = element.querySelector(
      ".dropdown-content li:nth-child(4) a"
    );
    fireEvent.click(upvotes);
    expect(onSortSpy).toHaveBeenCalledWith("");
  });

  test("renders submitted on option as active element", () => {
    component = render(
      <SortMenu sortBy="submittedOn" onSort={onSortSpy} orderAsc={true} />,
      {
        container: document.body.appendChild(element),
      }
    );
    const submittedOn = element.querySelector(
      ".dropdown-content li:nth-child(1)"
    );

    expect(submittedOn.classList.contains("active")).toBe(true);
  });

  test("renders sort arrows with submitted on option if current active sort by is submitted on", () => {
    component = render(
      <SortMenu sortBy="submittedOn" onSort={onSortSpy} orderAsc={true} />,
      {
        container: document.body.appendChild(element),
      }
    );
    let submittedOn = element.querySelector(".dropdown-content li:nth-child(1)");

    expect(submittedOn.innerHTML).toBe(
      `<a><i class=\"material-icons\">access_time</i>By Created On<i class=\"material-icons\">arrow_upward</i></a>`
    );

    component = render(
      <SortMenu sortBy="submittedOn" onSort={onSortSpy} orderAsc={false} />,
      {
        container: document.body.appendChild(element),
      }
    );
    submittedOn = element.querySelector(".dropdown-content li:nth-child(1)");

    expect(submittedOn.innerHTML).toBe(
      `<a><i class=\"material-icons\">access_time</i>By Created On<i class=\"material-icons\">arrow_downward</i></a>`
    );
  });

  test("renders upvotes option as active element", () => {
    component = render(
      <SortMenu sortBy="upvotes" onSort={onSortSpy} orderAsc={true} />,
      {
        container: document.body.appendChild(element),
      }
    );
    const upvotes = element.querySelector(".dropdown-content li:nth-child(2)");

    expect(upvotes.classList.contains("active")).toBe(true);
  });

  test("renders sort arrows with upvotes if current active sort by is upvotes", () => {
    component = render(
      <SortMenu sortBy="upvotes" onSort={onSortSpy} orderAsc={true} />,
      {
        container: document.body.appendChild(element),
      }
    );
    let upvotes = element.querySelector(".dropdown-content li:nth-child(2)");

    expect(upvotes.innerHTML).toBe(
      `<a><i class=\"material-icons\">thumb_up</i>By Upvotes<i class=\"material-icons\">arrow_downward</i></a>`
    );

    component = render(
      <SortMenu sortBy="upvotes" onSort={onSortSpy} orderAsc={false} />,
      {
        container: document.body.appendChild(element),
      }
    );
    upvotes = element.querySelector(".dropdown-content li:nth-child(2)");

    expect(upvotes.innerHTML).toBe(
      `<a><i class=\"material-icons\">thumb_up</i>By Upvotes<i class=\"material-icons\">arrow_upward</i></a>`
    );
  });
});
