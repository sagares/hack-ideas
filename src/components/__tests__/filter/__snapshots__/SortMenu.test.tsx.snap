// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Sort Menu Component renders 1`] = `
<div>
  <a
    class="dropdown-trigger btn-floating btn-large"
    data-target="dropdown1"
  >
    <i
      class="large material-icons"
    >
      sort
    </i>
  </a>
  <ul
    class="dropdown-content"
    id="dropdown1"
    tabindex="0"
  >
    <li
      class=""
      tabindex="0"
    >
      <a>
        <i
          class="material-icons"
        >
          access_time
        </i>
        By Created On
      </a>
    </li>
    <li
      class=""
      tabindex="0"
    >
      <a>
        <i
          class="material-icons"
        >
          thumb_up
        </i>
        By Upvotes
      </a>
    </li>
    <li
      class="divider"
      tabindex="-1"
    />
    <li
      tabindex="0"
    >
      <a>
        <i
          class="material-icons"
        >
          cached
        </i>
        Reset
      </a>
    </li>
  </ul>
</div>
`;
